p2p/src/authenticated/actors/tracker/actor.rs

```diff
@@ -398,9 +392,20 @@ impl<E: Spawner + Rng + Clock + GClock + Metrics, C: Scheme> Actor<E, C> {
         Ok(Some(peers))
     }
 
+    /// Attempt to reserve a connection to a peer, returning a [`Reservation`] if successful.
+    ///
+    /// Will return `None` in any of the following cases:
+    /// - The peer is already reserved.
+    /// - The peer is not [`Self::allowed`].
+    /// - The peer has been rate-limited for connection attempts.
     fn reserve(&mut self, peer: C::PublicKey) -> Option<Reservation<E, C>> {
         // Check if we are already reserved
-        if self.connections.contains(&peer) {
+        if self.reserved.contains(&peer) {
+            return None;
+        }
+
+        // Check if peer is invalid or blocked
+        if !self.allowed(&peer) {
```

Comment:
Was not checking this a bug for incoming connections on `main`?

Seems like you could have:
1) `handshake`: https://github.com/commonwarexyz/monorepo/blob/103caa03ef469272b29f9f89a35bf27bd2d8f8a9/p2p/src/authenticated/actors/listener.rs#L66
2) `reserve`: https://github.com/commonwarexyz/monorepo/blob/103caa03ef469272b29f9f89a35bf27bd2d8f8a9/p2p/src/authenticated/actors/listener.rs#L99
3) `spawn`: https://github.com/commonwarexyz/monorepo/blob/103caa03ef469272b29f9f89a35bf27bd2d8f8a9/p2p/src/authenticated/actors/listener.rs#L129

(where `construct` would "kill" but this would still be suboptimal to spawn a peer we don't care about)

Reply:
Yes I think it was a bug (not disastrous but unintended)

Reply:
Let's add a regression test for this (using `stream` I think?)

Reply:
https://github.com/commonwarexyz/monorepo/pull/908/files#diff-44e3c04630b5ce442d9acd5bc69792ede1d18306955f4652d4842e95675db41bR1049

Test here to make sure we don't give out reservations for unknown peers